File: src/ios/WebSocketClientManager.m
//
//  WebSocketClientManager.m
//  WebSocket
//
//  Created by Beatriz on 23/05/2024.
//

#import <Foundation/Foundation.h>
#import <Combine/Combine.h>
#import "WebSocketClient.m"

#import "WebSocketClientManager.h"

@implementation WebSocketClientManager

- (instancetype)init {
    self = [super init];
    
    if (self) {
        _webSocketClient = [[WebSocketClient alloc] init];
        _isConnected = NO;
        _sendAPDUMessageCount = 0;
        _cancellables = [NSMutableSet set];
        [self subscribeToConnectionStatus];
        [self subscribeToWebSocketMessages];
    }

    return self;
}

- (void)connectToURL:(NSString *)url {
    [self.webSocketClient connectToWebSocketUrl:url];
}

- (void)sendJSONObject:(id)jsonObject onSuccess:(void (^)(void))onSuccess {
    [self.webSocketClient send:jsonObject onSuccess:onSuccess];
}

- (void)sendMessage:(NSString *)message onSuccess:(void (^)(void))onSuccess {
    [self.webSocketClient.socket writeString:message];
    onSuccess();
}

- (void)subscribeToConnectionStatus {
    [self.webSocketClient.connectionStatusPublisher
        sinkWithWeakTarget:self weakTargetAction:@selector(updateConnectionStatus:)]
        .storeIn:self.cancellables];
}

- (void)updateConnectionStatus:(NSNumber *)isConnected {
    self.isConnected = isConnected.boolValue;
}

- (void)subscribeToWebSocketMessages {
    [self.webSocketClient.messagePublisher
        sinkWithWeakTarget:self weakTargetAction:@selector(handleWebSocketMessage:)]
        .storeIn:self.cancellables];
}

- (void)handleWebSocketMessage:(NSString *)message {
    NSLog(@"Handling WebSocket message: %@", message);
    NSData *data = [message dataUsingEncoding:NSUTF8StringEncoding];
    
    if (!data) {
        NSLog(@"Failed to convert message to data.");
        return;
    }
    
    NSError *error = nil;
    NSArray *jsonArray = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
    
    if (error || ![jsonArray isKindOfClass:[NSArray class]] || jsonArray.count <= 2) {
        NSLog(@"Failed to parse JSON with error: %@", error);
        return;
    }
    
    NSDictionary *messageDict = jsonArray[0];
    NSString *messageType = messageDict[@"type"];
    NSString *correlationId = (jsonArray[2] == [NSNull null]) ? @"" : jsonArray[2];
    
    if ([messageType isEqualToString:@"confirmSMSCodeResponse"]) {
        NSString *payload = messageDict[@"payload"];
        NSLog(@"Received confirmSMSCodeResponse message with payload: %@", payload);
        [[NSNotificationCenter defaultCenter] postNotificationName:@"confirmSMSCodeResponse" object:@{@"payload": payload}];
    } else if ([messageType isEqualToString:@"sendAPDU"]) {
        NSString *payload = messageDict[@"payload"];
        NSLog(@"Received sendAPDU message with payload: %@ and correlationId: %@", payload, correlationId);
        
        self.sendAPDUMessageCount += 1;
        
        if (self.sendAPDUMessageCount == 1) {
            [[NSNotificationCenter defaultCenter] postNotificationName:@"receivedFirstSendAPDU" object:@{@"payload": payload, @"correlationId": correlationId}];
        } else if (self.sendAPDUMessageCount == 2) {
            [[NSNotificationCenter defaultCenter] postNotificationName:@"receivedSecondSendAPDU" object:@{@"payload": payload, @"correlationId": correlationId}];
        }
    } else if ([messageType isEqualToString:@"eRezeptTokensFromAVS"]) {
        NSLog(@"Received eRezeptTokensFromAVS message with correlationId: %@", correlationId);
        NSString *payload = messageDict[@"payload"];
        [[NSNotificationCenter defaultCenter] postNotificationName:@"receivedERezeptTokensFromAVS" object:@{@"payload": payload, @"correlationId": correlationId}];
    } else if ([messageType isEqualToString:@"eRezeptBundlesFromAVS"]) {
        NSLog(@"Received eRezeptBundlesFromAVS message with correlationId: %@", correlationId);
        NSString *payload = messageDict[@"payload"];
        [[NSNotificationCenter defaultCenter] postNotificationName:@"receivedERezeptBundlesFromAVS" object:@{@"payload": payload, @"correlationId": correlationId}];
    } else {
        NSLog(@"Other message type: %@", messageType);
    }
}

@end

@implementation MockWebSocketClientManager

- (instancetype)init {
    self = [super init];
    if (self) {
        self.isConnected = NO;
    }
    return self;
}

- (void)connectToURL:(NSString *)url {
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        self.isConnected = YES;
    });
}

@end

File: src/ios/WebSocketClientManager.swift
//
//  WebSocketClientManager.swift
//  WebSocket
//
//  Created by Beatriz on 23/05/2024.
//

import Combine
import Foundation

public class WebSocketClientManager: ObservableObject {
    @Published public var webSocketClient: WebSocketClient?
    @Published public var isConnected = false
    @Published public var cardSessionId: String?
    public var cancellables = Set<AnyCancellable>()
    private var sendAPDUMessageCount = 0

    public init() {
        self.webSocketClient = WebSocketClient()
        subscribeToConnectionStatus()
        subscribeToWebSocketMessages()
    }

    public func connect(to url: String) {
        webSocketClient?.connect(to: url)
    }

    public func send(_ jsonObject: Any, onSuccess: @escaping () -> Void) {
        webSocketClient?.send(jsonObject, onSuccess: onSuccess)
    }

    public func send(_ message: String, onSuccess: @escaping () -> Void) {
        webSocketClient?.socket.write(string: message)
        onSuccess()
    }

    private func subscribeToConnectionStatus() {
        webSocketClient?.connectionStatusPublisher
            .sink { [weak self] isConnected in
                self?.isConnected = isConnected
            }
            .store(in: &cancellables)
    }

    private func subscribeToWebSocketMessages() {
        webSocketClient?.messagePublisher
            .sink { [weak self] message in
                self?.handleWebSocketMessage(message)
            }
            .store(in: &cancellables)
    }

    private func handleWebSocketMessage(_ message: String) {
        print("Handling WebSocket message: \(message)")
        guard let data = message.data(using: .utf8) else {
            print("Failed to convert message to data.")
            return
        }

        do {
            if let jsonArray = try JSONSerialization.jsonObject(
                with: data,
                options: []
            ) as? [Any],
                jsonArray.count > 2,
                let messageDict = jsonArray[0] as? [String: Any],
                let messageType = messageDict["type"] as? String,
                let correlationId: String? = {
                    if jsonArray[2] is NSNull {
                        return ""
                    } else {
                        return jsonArray[2] as? String
                    }
                }()
            {

                switch messageType {

                case "confirmSMSCodeResponse":
                    if let payload = messageDict["payload"] as? String {
                        print(
                            "Received confirmSMSCodeResponse message with payload: \(payload)"
                        )
                        NotificationCenter.default.post(
                            name: .confirmSMSCodeResponse,
                            object: ["payload": payload]
                        )
                    }

                case "sendAPDU":
                    if let payload = messageDict["payload"] as? String {
                        if let correlationId = correlationId {
                            print(
                                "Received sendAPDU message with payload: \(payload), correlationId: \(correlationId)"
                            )
                        } else {
                            print(
                                "Received sendAPDU message with payload: \(payload), correlationId: nil"
                            )
                        }
                        NotificationCenter.default.post(
                            name: .receivedSendAPDU,
                            object: [
                                "payload": payload,
                                "correlationId": correlationId,
                            ]
                        )
                    } else {
                        print(
                            "Received sendAPDU message but payload is missing or invalid."
                        )
                    }

                case "eRezeptTokensFromAVS":
                    if let correlationId = correlationId {
                        print(
                            "Received eRezeptTokensFromAVS message with correlationId: \(correlationId)"
                        )
                    } else {
                        print(
                            "Received eRezeptTokensFromAVS message with no correlationId"
                        )
                    }
                    if let payload = messageDict["payload"] as? String {
                        NotificationCenter.default.post(
                            name: .receivedERezeptTokensFromAVS,
                            object: [
                                "payload": payload,
                                "correlationId": correlationId,
                            ]
                        )
                    }

                case "eRezeptBundlesFromAVS":
                    if let correlationId = correlationId {
                        print(
                            "Received eRezeptBundlesFromAVS message with correlationId: \(correlationId)"
                        )
                    } else {
                        print(
                            "Received eRezeptBundlesFromAVS message with no correlationId"
                        )
                    }
                    if let payload = messageDict["payload"] as? String {
                        NotificationCenter.default.post(
                            name: .receivedERezeptBundlesFromAVS,
                            object: [
                                "payload": payload,
                                "correlationId": correlationId,
                            ]
                        )
                    }
                    
                case "receiveTasklistError":
                    print("Received receiveTasklistError message with correlationId: \(correlationId)")
                    if let payload = messageDict["payload"] as? String {
                        NotificationCenter.default.post(name: .receivedTasklistError, object: ["payload": payload, "correlationId": correlationId])
                    }

                default:
                    print("Other message type: \(messageType)")
                }
            }
        } catch {
            print("Failed to parse JSON with error: \(error)")
        }
    }
}

extension Notification.Name {
    public static let confirmSMSCodeResponse = Notification.Name(
        "confirmSMSCodeResponse"
    )
    public static let receivedSendAPDU = Notification.Name("receivedSendAPDU")
    public static let receivedERezeptTokensFromAVS = Notification.Name(
        "receivedERezeptTokensFromAVS"
    )
    public static let receivedERezeptBundlesFromAVS = Notification.Name(
        "receivedERezeptBundlesFromAVS"
    )
    static let receivedTasklistError = Notification.Name("receivedTasklistError")
}

public class MockWebSocketClientManager: WebSocketClientManager {
    public override init() {
        super.init()
        self.isConnected = false
    }

    public override func connect(to url: String) {
        DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
            self.isConnected = true
        }
    }
}
